File: m4-h2/src/test/java/ru/yegorpilipenko/otus/spring/m4h2/service/BookServiceTest.java
package ru.yegorpilipenko.otus.spring.m4h2.service;

import lombok.val;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import ru.yegorpilipenko.otus.spring.m4h2.model.Book;
import ru.yegorpilipenko.otus.spring.m4h2.model.Comment;
import ru.yegorpilipenko.otus.spring.m4h2.repository.BookRepository;

import java.util.List;
import java.util.Optional;

import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

@ExtendWith(MockitoExtension.class)
class BookServiceTest implements WithAssertions {

    @Mock
    BookRepository bookRepository;
    BookService bookService;

    @BeforeEach
    void setup() {
        bookService = new BookServiceImpl(bookRepository);
    }

    @Nested
    class RemoveByIdTest {

        @Test
        void shouldPassIfExists() {
            val id = "1";
            doReturn(1).when(bookRepository).removeById(id);
            assertThat(bookService.removeById(id)).isTrue();
            verify(bookRepository).removeById(id);
            verifyNoMoreInteractions(bookRepository);
        }

        @Test
        void shouldPassIfNotExists() {
            val id = "1";
            doReturn(0).when(bookRepository).removeById(id);
            assertThat(bookService.removeById(id)).isFalse();
            verify(bookRepository).removeById(id);
            verifyNoMoreInteractions(bookRepository);
        }

    }

    @Nested
    class AddCommentTest {

        @Test
        void shouldPassForNormalBook() {
            val bookId = "1";
            val comment = new Comment("e2", "t2");
            val book = Book.builder()
                    .id(bookId)
                    .name("n")
                    .comments(List.of(new Comment("e1", "t1")))
                    .build();
            val expected = Book.builder()
                    .id(bookId)
                    .name("n")
                    .comments(List.of(new Comment("e1", "t1"), new Comment("e2", "t2")))
                    .build();
            doReturn(Optional.of(book)).when(bookRepository).findById(bookId);
            doReturn(expected).when(bookRepository).save(expected);
            assertThat(bookService.addComment(bookId, comment)).isEqualTo(expected);
            verify(bookRepository).findById(bookId);
            verify(bookRepository).save(expected);
            verifyNoMoreInteractions(bookRepository);
        }

        @Test
        void shouldFailIfBookNotExists() {
            val bookId = "1";
            doReturn(Optional.empty()).when(bookRepository).findById(bookId);
            assertThatThrownBy(() -> bookService.addComment(bookId, new Comment("e", "t")))
                    .isInstanceOf(IllegalArgumentException.class);
            verify(bookRepository).findById(bookId);
            verifyNoMoreInteractions(bookRepository);
        }

    }

    @Nested
    class RemoveCommentTest {

        @Test
        void shouldPassForExistsComment() {
            val bookId = "1";
            val comment = new Comment("e", "t");
            val book = Book.builder()
                    .id(bookId)
                    .name("n")
                    .comments(List.of(comment))
                    .build();
            val expected = Book.builder()
                    .id(bookId)
                    .name("n")
                    .build();
            doReturn(Optional.of(book)).when(bookRepository).findById(bookId);
            doReturn(expected).when(bookRepository).save(expected);
            assertThat(bookService.removeComment(bookId, comment)).get().isEqualTo(expected);
            verify(bookRepository).findById(bookId);
            verify(bookRepository).save(expected);
            verifyNoMoreInteractions(bookRepository);
        }

        @Test
        void shouldPassForNotExistsComment() {
            val bookId = "1";
            val comment = new Comment("e2", "t2");
            val book = Book.builder()
                    .id(bookId)
                    .name("n")
                    .comments(List.of(new Comment("e1", "t1")))
                    .build();
            doReturn(Optional.of(book)).when(bookRepository).findById(bookId);
            assertThat(bookService.removeComment(bookId, comment)).isEmpty();
            verify(bookRepository).findById(bookId);
            verifyNoMoreInteractions(bookRepository);
        }

        @Test
        void shouldFailIfBookNotExists() {
            val bookId = "1";
            doReturn(Optional.empty()).when(bookRepository).findById(bookId);
            assertThatThrownBy(() -> bookService.removeComment(bookId, new Comment("e", "t")))
                    .isInstanceOf(IllegalArgumentException.class);
            verify(bookRepository).findById(bookId);
            verifyNoMoreInteractions(bookRepository);
        }

    }

}

File: m2-h3/src/test/kotlin/ru/yegorpilipenko/otus/spring/m2h3/service/BookServiceTest.kt
package ru.yegorpilipenko.otus.spring.m2h3.service

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.assertj.core.api.WithAssertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.junit.jupiter.MockitoExtension
import ru.yegorpilipenko.otus.spring.m2h3.model.Book
import ru.yegorpilipenko.otus.spring.m2h3.model.Comment
import ru.yegorpilipenko.otus.spring.m2h3.repository.BookRepository
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class BookServiceTest: WithAssertions {

    @Mock
    lateinit var bookRepository: BookRepository
    lateinit var bookService: BookService

    @BeforeEach
    fun setup() {
        bookService = BookServiceImpl(bookRepository)
    }

    @Nested
    inner class RemoveByIdTest {

        @Test
        fun shouldPassForExists() {
            doReturn(1).whenever(bookRepository).removeById(any())
            assertThat(bookService.removeById("1")).isTrue()
            verify(bookRepository).removeById("1")
            verifyNoMoreInteractions(bookRepository)
        }

        @Test
        fun shouldPassForNotExists() {
            assertThat(bookService.removeById("1")).isFalse()
            verify(bookRepository).removeById("1")
            verifyNoMoreInteractions(bookRepository)
        }

    }

    @Nested
    inner class AddCommentTest {

        @Test
        fun shouldPassForExistsBook() {
            val bookId = "1"
            doReturn(Optional.of(Book(
                    id = bookId,
                    name = "testBook",
                    comments = listOf(Comment("e1", "t1"), Comment("e2", "t2"))
            ))).whenever(bookRepository).findById(bookId)
            doReturn(Book(name = "stub")).whenever(bookRepository).save(any<Book>())
            assertThat(bookService.addComment(bookId, Comment("e3", "t3"))).isEqualTo(Book(name = "stub"))
            verify(bookRepository).findById(bookId)
            verify(bookRepository).save(Book(
                    id = bookId,
                    name = "testBook",
                    comments = listOf(Comment("e1", "t1"), Comment("e2", "t2"), Comment("e3", "t3"))
            ))
            verifyNoMoreInteractions(bookRepository)
        }

        @Test
        fun shouldFailForNotExistsBook() {
            val bookId = "1"
            doReturn(Optional.empty<Book>()).whenever(bookRepository).findById(bookId)
            assertThatThrownBy { bookService.addComment(bookId, Comment("e1", "t1")) }
                    .isInstanceOf(IllegalArgumentException::class.java)
            verify(bookRepository).findById(bookId)
            verifyNoMoreInteractions(bookRepository)
        }

    }

    @Nested
    inner class RemoveCommentTest {

        @Test
        fun shouldPassForExistsComment() {
            val bookId = "1"
            doReturn(Optional.of(Book(
                    id = bookId,
                    name = "testBook",
                    comments = listOf(Comment("e1", "t1"), Comment("e2", "t2"))
            ))).whenever(bookRepository).findById(bookId)
            doReturn(Book(name = "stub")).whenever(bookRepository).save(any<Book>())
            assertThat(bookService.removeComment(bookId, Comment("e1", "t1"))).isEqualTo(Book(name = "stub"))
            verify(bookRepository).findById(bookId)
            verify(bookRepository).save(Book(
                    id = bookId,
                    name = "testBook",
                    comments = listOf(Comment("e2", "t2"))
            ))
            verifyNoMoreInteractions(bookRepository)
        }

        @Test
        fun shouldPassForNotExistsComment() {
            val bookId = "1"
            doReturn(Optional.of(Book(
                    id = bookId,
                    name = "testBook",
                    comments = listOf(Comment("e1", "t1"), Comment("e2", "t2"))
            ))).whenever(bookRepository).findById(bookId)
            assertThat(bookService.removeComment(bookId, Comment("e3", "t3"))).isNull()
            verify(bookRepository).findById(bookId)
            verifyNoMoreInteractions(bookRepository)
        }

        @Test
        fun shouldFailForNotExistsBook() {
            val bookId = "1"
            doReturn(Optional.empty<Book>()).whenever(bookRepository).findById(bookId)
            assertThatThrownBy { bookService.removeComment(bookId, Comment("e1", "t1")) }
                    .isInstanceOf(IllegalArgumentException::class.java)
            verify(bookRepository).findById(bookId)
            verifyNoMoreInteractions(bookRepository)
        }

    }

}